rpc/handler.go

```diff
@@ -0,0 +1,383 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/gin-gonic/gin"
+	"github.com/goccy/go-json"
+	"github.com/mitchellh/mapstructure"
+	"github.com/puzpuzpuz/xsync/v3"
+	"github.com/stackup-wallet/stackup-bundler/pkg/client"
+	"github.com/stackup-wallet/stackup-bundler/pkg/jsonrpc"
+
+	"github.com/blndgs/bundler/srv"
+)
+
+// ExtERC4337Controller extends the default JSON-RPC controller to handle non-ERC4337 Ethereum RPC methods.
+func ExtERC4337Controller(hashesMap *xsync.MapOf[string, srv.OpHashes], rpcAdapter *client.RpcAdapter, rpcClient *rpc.Client, ethRPCClient *ethclient.Client) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if c.Request.Method != "POST" {
+			jsonrpcError(c, -32700, "Parse error", "POST method excepted", nil)
+			return
+		}
+
+		if c.Request.Body == nil {
+			jsonrpcError(c, -32700, "Parse error", "No POST data", nil)
+			return
+		}
+
+		body, err := readBody(c)
+		if err != nil {
+			jsonrpcError(c, -32700, "Parse error", "Error while reading request body", nil)
+			return
+		}
+
+		data := make(map[string]any)
+		err = json.Unmarshal(body, &data)
+		if err != nil {
+			jsonrpcError(c, -32700, "Parse error", "Error parsing json request", nil)
+			return
+		}
+
+		id, ok := parseRequestId(data)
+		if !ok {
+			jsonrpcError(c, -32600, "Invalid Request", "No or invalid 'id' in request", nil)
+			return
+		}
+
+		if data["jsonrpc"] != "2.0" {
+			jsonrpcError(c, -32600, "Invalid Request", "Version of jsonrpc is not 2.0", &id)
+			return
+		}
+
+		method, ok := data["method"].(string)
+		if !ok {
+			jsonrpcError(c, -32600, "Invalid Request", "No or invalid 'method' in request", &id)
+			return
+		}
```

Comment:
this boilerplate code runs twice for rpc methods handled by the stackup controller. If possible, it would be nice to run once only.